### PR TITLE
Noti system #73

### PR DESCRIPTION
**Pull request by pospieszynski**
Notifications system, feedback pls;]

### CODE REVIEW

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def update
   end
 
   def destroy
-    @event.destroy
+    @event.destroy_and_notify current_user, :projects_contributors
```
Comment (tprochniewicz) [inline app/controllers/events_controller.rb]: add brackets

File: app/controllers/notifications_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  def index
+    @notifications = current_user.incoming_notifications.not_read
+  end
+
+  def mark_as_read
+    @notification = Notification.find(params[:id])
+    if @notification.update_attribute(:read, true)
+      redirect_to notifications_path
+    else
+      redirect_to root_path, notice: "Unable to update notificaion"
```
Comment (tprochniewicz) [inline app/controllers/notifications_controller.rb]: typo

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -22,7 +22,7 @@ def edit
   def create
     @event = @project.events.new(event_params)
     @event.author = current_user
-    if @event.save
+    if @event.save_and_notify(current_user, :projects_contributors )
```
Comment (filip373) [inline app/controllers/events_controller.rb]: you can remove empty space before ')'

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -48,7 +48,7 @@ def mark_as_done
   end
 
   def destroy
-    @task.destroy
+    @task.destroy_and_notify current_user, :participants
```
Comment (tprochniewicz) [inline app/controllers/tasks_controller.rb]: brackets

File: app/models/concerns/notification_methods.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    end
+  end
+
+  def update_and_notify(params, current_user, target)
+    transaction do
+      begin
+        self.update!(params)
+        attributes_changed = self.previous_changes
+        attributes_changed.delete("updated_at")
+        puts attributes_changed.to_s
```
Comment (tprochniewicz) [inline app/models/concerns/notification_methods.rb]: puts

File: spec/models/event_spec.rb [hunk 6]
```diff
@@ -36,7 +36,7 @@
     it "adds error message about author" do
       subject.validate
       expect(subject.errors.messages[:author].first).
-        to eql("must take part in project")
+          to eql("must take part in project")
```
Comment (tprochniewicz) [inline spec/models/event_spec.rb]: not needed

File: spec/factories/notifications.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
```
Comment (tprochniewicz) [inline spec/factories/notifications.rb]: not needed

File: spec/models/event_spec.rb [hunk 8]
```diff
@@ -71,4 +71,105 @@
...
+    let(:another_user) { FactoryGirl.create(:user) }
+    let(:different_user) { FactoryGirl.create(:user) }
+    let(:message) { "#{event.class} - #{event.title} has been recently updated :* description was changed to wind of change" }
+
+    before do
+      UserProject.create(user: another_user, project: event.project)
+      UserProject.create(user: different_user, project: event.project)
+      params = event.attributes
+      params["description"] = "wind of change"
+      event.update_and_notify params, user, :projects_contributors
```
Comment (tprochniewicz) [inline spec/models/event_spec.rb]: brackets

File: spec/models/event_spec.rb [hunk 9]
```diff
@@ -71,4 +71,105 @@
...
+    end
+
+    it "updates event record" do
+      expect(event.description).to eql("wind of change")
+    end
+
+    it "sends notification to users that contribute to project it terms of which events exist" do
+
```
Comment (tprochniewicz) [inline spec/models/event_spec.rb]: drop line

File: spec/models/project_spec.rb [hunk 10]
```diff
@@ -111,4 +111,90 @@
...
+    let(:project) { FactoryGirl.create(:project) }
+    let(:author) { project.user}
+    let(:user) { FactoryGirl.create(:user) }
+    let(:message) { "#{project.class}  - #{project.title} has been removed." }
+
+    before do
+      UserProject.create(user: user, project: project)
+      project.destroy_and_notify author, :contributors
+    end
+
```
Comment (tprochniewicz) [inline spec/models/project_spec.rb]: drop line

File: spec/models/task_spec.rb [hunk 11]
```diff
@@ -57,4 +57,95 @@
...
+
+    before do
+      UserProject.create(user: another_user, project: task.project)
+      UserProject.create(user: user, project: task.project)
+      UserTask.create(user: another_user, task: task)
+      task.destroy_and_notify user, :participants
+    end
+
+    it "deletes task form db" do
+      expect(Task.all).to eq([])
```
Comment (tprochniewicz) [inline spec/models/task_spec.rb]: be empty

**Comment by tprochniewicz** [comment]
:+1:  @gotar

File: app/models/concerns/notification_methods.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
+module NotificationMethods
+
```
Comment (gotar) [inline app/models/concerns/notification_methods.rb]: drop empty line

Comment (pospieszynski) [inline app/models/concerns/notification_methods.rb]: done

**Comment by gotar** [comment]
+1

**Pull request merged**